### ISSUE TITLE
MS SQL Server empty array inserts raises syntax error

### ISSUE THREAD

**Issue by Aiden0** [hasura/graphql-engine#8959]
### Version Information
Server Version: v2.10.0.cli-migrations-v3

### Environment
OSS

### What is the current behaviour?
If i have a simple table, in sql server database, with columns id and name and I insert an empty array like so:

```
mutation {
  insert_test_table(objects: []) {
    affected_rows
  }
}
```
 I get the error back:
 
 ```
 {
  "errors": [
    {
      "extensions": {
        "code": "unexpected",
        "internal": {
          "exception": {
            "message": "[Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Incorrect syntax near 'VALUES'.",
            "type": "unsuccessful_return_code"
          },
          "query": "INSERT INTO [dbo].[test_table] OUTPUT INSERTED.[name], INSERTED.[id] INTO #inserted([name], [id]) VALUES "
        },
        "path": "$"
      },
      "message": "database query error"
    }
  ]
}
 ```

### What is the expected behaviour?

Should give the response:

```
{
  "data": {
    "insert_test_tables": {
      "affected_rows": 0
    }
  }
}
```

which it does for  PostgreSQL database. 

### How to reproduce the issue?

1. Create any table
2. do graphql insert with objects being empty array on that table

### Can you identify the location in the source code where the problem exists?

I think it is occurring [here](https://github.com/hasura/graphql-engine/blob/master/server/src-lib/Hasura/Backends/MSSQL/Execute/Insert.hs#L199). I am not too familiar with Haskell, so will have to look at more later.

**Comment by Aiden0** [comment]
Hey @adas98012 so I think I get the same error in v2.10.0 when I add the object as only `{}`.  I upgraded to v2.11.2 which when doing `mutation {
  insert_States(objects: {}) {
    affected_rows
  }
}` it returns error:

```
{
  {
  "errors": [
    {
      "extensions": {
        "code": "unexpected",
        "internal": {
          "exception": {
            "message": "[Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Column name or number of supplied values does not match table definition.",
            "type": "unsuccessful_return_code"
          },
          "query": "INSERT INTO [dbo].[States] OUTPUT INSERTED.[name], INSERTED.[id], INSERTED.[country_id] INTO #inserted([name], [id], [country_id]) VALUES (DEFAULT)"
        },
        "path": "$"
      },
      "message": "database query error"
    }
  ]
}
``` 

which makes sense

```

Interestingly though if I have a table with one nullable column and a primary key, for example UserStatus with column id(int, primary key, unique) and description(nvarchar, nullable):

then 
```
mutation {
  insert_UserStatus(objects: {}) {
    affected_rows
  }
}
```
it adds the data which yields:
```
{
  "data": {
    "insert_UserStatus": {
      "affected_rows": 1
    }
  }
}
```
### However when I add  empty array`[]` instead
I get same error I posted before regardless of table:
![image](https://user-images.githubusercontent.com/26369669/190287739-39f56364-203f-4e7e-b71b-db6c702e9628.png)
 
If I add `[{}]` instead then I get the same response as adding just `{}` which makes sense.

**Comment by adas98012** [comment]
@Aiden0 Thanks for trying out all the scenarios. Looks like we have a viable workaround to use "[{}]" for an empty array. 
Please close the ticket if you are ok with the workaround.

**Comment by Aiden0** [comment]
Hey @adas98012 but I want to send an empty array not an array of malformed objects. Thus I except the result for handling an empty array i.e `[]` would be similar to using the PostgresSql driver. Which would be:

```
{
  "data": {
    "insert_test_tables": {
      "affected_rows": 0
    }
  }
}

```
, not incorrect syntax.

Otherwise I have to check the length of the array is 0 every time and choose not to send the data. Which usually wouldn't be so bad, but I can't do that if I want to group my mutations together in one request.

For example I want to delete all the corresponding rows and then reinsert new ones in single request, this will would lead to a fail and roll it back.

**Comment by lucasnad27** [comment]
I'm facing similar difficulties when inserting an empty array with SQL Server. The workaround is reasonable, but it would be great to have a proper fix to match parity with Postgres driver :)

**Comment by lucasnad27** [comment]
Correction, this workaround does not work for my use case. I'm still getting an error.

Payload looks something like this:
```json
{
  "query": "...",
  "variables": {
    "foo": [
      {}
    ]
  }
}
```

`[Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Column name or number of supplied values does not match table definition.`

Without a reasonable workaround, I'm forced to chop this up into separate graphql calls 👎🏼 Perhaps I can tweak my table schema until this is patched?

**Comment by manasag** [comment]
This is now fixed in v2.29. We now check for an insert with no values and skip the insert query altogether if this is the case. See this [commit](https://github.com/hasura/graphql-engine/commit/98fdeda54e5d49c0dc2144718a80960ce978b3d6)
